### DISCUSSION TITLE
Migration from old signals to new signals - trains don't work

### DISCUSSION THREAD

**Discussion by raven2cz** [Opinionated-Blueprints/10-Books-Full-of-Rails#96] [Q&A]
### Have you checked the following resources?

- [X] [Wiki](https://github.com/Opinionated-Blueprints/10-Books-Full-of-Rails/wiki)
- [X] [Discussions Forum](https://github.com/Opinionated-Blueprints/10-Books-Full-of-Rails/discussions)
- [X] [Open and Closed Issues](https://github.com/Opinionated-Blueprints/10-Books-Full-of-Rails/issues)

### Version of the Book

v2.21.2

### Bug Description

Migration from old signals to new signals - trains don't work

### How to reproduce the problem

I really wanted to figure this out, but I've already spent two days on it and still nothing. I need your advice. I was using an older version of your blueprints in my project, and I updated to the new version, but I can't get the new signaling system to work after the redesign.

In the new version, the trains are setting the provider stop incorrectly; they don't leave the variable there. Then they just sit at the station doing nothing. When I used the old trains, they worked fine.

I even tried switching to the new stations, but the result is still the same. The train says it can't find a route and stops because the specified provider stop doesn't exist.

I've saved my current setup for you. I'm using both the new approach and the old approach.

I really don't know what to do anymore, and I can't solve this on my own. :(((

My saves with problem is here (I added blueprints too, placed in onedrive):
https://1drv.ms/f/s!Aq_X66v0FnfLnpszZd0JJxEToIal3Q?e=3pu8pu

### What did you expect?

The trains will find the correct provider station.

### What happened instead?

The trains have defined wrong provider station which doesn't exist.

**Comment by Prorok2017** [comment]
@raven2cz Your train is missing a stopping point - an unloading station.
![image](https://github.com/user-attachments/assets/fc7a9db2-a50e-44bf-81d8-701b9703dffc)
![image](https://github.com/user-attachments/assets/3f8848f0-e872-4987-a119-ad24954c0f4f)

in order for a train to depart to a new loading station, the train must be completely empty.

I corrected your save - added the station to the list of interruptions
https://drive.google.com/file/d/1YeImKrGqYhu8zeuM5rg4ceAtXlZnvEmd/view?usp=sharing

**Reply by raven2cz** [reply]
Thank you for identifying the problem. However, the fix might be a bit more challenging. When I deleted the blueprint and imported it again, the exception still wasn’t there. This means the save somehow remembers not to add it again, right? So I won’t have to do it manually for every train? Does the save just decide to overwrite it and ignore the blueprint?

**Reply by raven2cz** [reply]
Why am I asking this? Because I have this on a shared server, and while your save is fixed, you include your mods that were saved with it. I cannot upload it to the server like this. I need to know how to fix it or get modified save without additional mods, just vanilla state.

**Reply by raven2cz** [reply]
@Prorok2017 is it possible to patch the save without your mods for vanilla server? Or quick guide how to create the generic exception. If I delete the blueprint and import it again, the exception still missing.

**Reply by Prorok2017** [reply]
@raven2cz  My mods do not affect trains, stations and logic network. You can fix trains without your own changes in old blueprints as follows - delete old blueprints -> import new -> delete all trains -> install from new blueprints. If after these steps it still does not start - the problem may be in the tracks - check for additional traffic lights - I accidentally put one traffic light inside a rail block and my track changed direction =D. Because of this, trains stopped finding tracks to stations

![image](https://github.com/user-attachments/assets/54096e2d-fbd5-440d-9d47-24bceceda051)

![image](https://github.com/user-attachments/assets/f1d40bbf-ec84-49db-8eb0-3db800a2abb1)

**Comment by brandon-gradient** [comment]
The interrupts (you call them "exceptions") are part of specific groups used by the blueprint. This allows editing an interrupt or stop for any of the trains in the group and it changes it for the rest of them in the group. This is a QoL so similar trains don't have to be changed 1 by 1 if they are in the same group.
What this means for your save is that the group used by the blueprint already exists in your save, and when using a train from this book, that train's group settings will conform to the already existing matching group.

My suggestion to you is to use a train blueprint from the book but when placing it, pick a group that doesn't already exist in your save, so not the "A" virtual signal (or others you already have), and

1. Place it down
2. Note the extra interrupt that is missing from your existing trains
3. Click the pencil as if you were going to edit it
4. Write down everything you see for replication in existing train groups
5. Recreate the interrupt in existing trains make changes where the sample interrupt uses it's group icon and substituting for existing train's group (it seems like "A" virtual signal for at least one of them)

I'm making some assumptions here that might not hold but if you're using the most recent version of this blueprint book this should work

**Reply by raven2cz** [reply]
> The interrupts (you call them "exceptions") are part of specific groups used by the blueprint. This allows editing an interrupt or stop for any of the trains in the group and it changes it for the rest of them in the group. This is a QoL so similar trains don't have to be changed 1 by 1 if they are in the same group. What this means for your save is that the group used by the blueprint already exists in your save, and when using a train from this book, that train's group settings will conform to the already existing matching group.
> 
> My suggestion to you is to use a train blueprint from the book but when placing it, pick a group that doesn't already exist in your save, so not the "A" virtual signal (or others you already have), and
> 
>     1. Place it down
> 
>     2. Note the extra interrupt that is missing from your existing trains
> 
>     3. Click the pencil as if you were going to edit it
> 
>     4. Write down everything you see for replication in existing train groups
> 
>     5. Recreate the interrupt in existing trains make changes where the sample interrupt uses it's group icon and substituting for existing train's group (it seems like "A" virtual signal for at least one of them)
> 
> 
> I'm making some assumptions here that might not hold but if you're using the most recent version of this blueprint book this should work

Thank you very much for your response and explanation. In the end, I forced myself to study everything thoroughly and arrived at the same procedure. However, after precisely replicating the interrupts into group A, while the trains started running correctly, after a cycle, instead of continuing to the next interrupt, they suddenly switched from automatic mode to manual mode and stopped. The interrupts are identical. I assume there might be some internal issue or bug in Factorio.

In the end, I completely removed all the trains, reloaded the cargo, and created a new train group 3 (since I use 1-2-0). The blueprints applied correctly when creating the group, the interrupts are set properly, and there is no switching between automatic and manual mode. Although it took about 3 hours to redo everything for the new group, it was worth the effort.

Once again, thank you so much for your help and especially for the excellent blueprint!
